Library/Homebrew/tap.rb

```diff
@@ -280,4 +316,93 @@ def self.each
   def self.names
     map(&:name)
   end
+
+  private
+
+  # @private
+  def formula_file_to_name(file)
+    "#{name}/#{file.basename(".rb")}"
+  end
+
+  # @private
+  def alias_file_to_name(file)
+    "#{name}/#{file.basename}"
+  end
+end
+
+class CoreFormulaTap < Tap
```

Comment:
> 1. We will delete Library/Formula, Library/Aliases and Library/Homebrew/formula_rename.rb in the core.

This makes me a bit nervous. I'm wondering if it's worth having the tap check out into those locations. There will likely be older tools around Homebrew than make assumptions about these paths (and we've said "the filesystem is our API" for a long time).

> 2. We will export formula files with full git log into a tap repo. In this process, an equivalent formula_rename.json will be created.

:+1:. I also wonder if it's worth considering moving extracting core into a new repository too; this repository is pretty massive at this point particularly if you pull all e.g. PR heads. This could be a later stage

> 3. This core formula tap will be tapped to Library/Taps/homebrew/homebrew. At this point, core formula tap is identical to normal tap. Except, it cannot be uninstalled and pinned/unpinned.

:+1: except I do think it should be able to be uninstalled (or at least another tap chosen somehow as the "primary" tap). This would be useful for Linuxbrew and other users of the Homebrew core code (which we can actually apply semver to and perhaps even turn into a RubyGem).

> 4. In brew.rb, a migration check will be added to ensure the new tap is installed.

:+1: given we support the custom taps as above (eventually; doesn't need to be in the first version).



> At first appearance, module seems to look better. But, its code will be pretty strange as it replies the methods outside formula i.e. from the Tap class to work.

I'm :ok: with this particularly if it's declared in the same file.



> I have no objection. But I need to remind you that we won't change this class name even after core formula separation in the future in order to keep backward compatibility. If you and other maintainers are OK this, I will go head to change it to that name.

I think it's definitely fine to rename internal classes that we mark as `@private`.